Approximate time: 20 minutes

## Goals
- Understand FASTQ file format
- Run FastQC to asses data quality

## Take a look at our raw data

### FASTQ format

FASTQ files is the most common way to store biological sequence data.
Depending on the sequencing protocol, a single FASTQ file can represent an entire flow cell, a single lane, a single sample, or a portion of a sample.
We have two FASTQ files in our `raw_data` folder, which are the paired end data of a single sample.

From our course directory `into-to-ngs` change into the `raw data` directory:
```markdown
cd raw_data
```

Use the command `head` to look at the first few lines of our first FASTQ file.

```markdown
head na12878_1.fq
```

Each read in our file is represented are by four lines: An identifier, the nucleotide sequence, an optional second identifier and a quality string.
Below is an example, the  arrows on the right show explanation of each line:

```bash
@SRR098401.109756285/1                   <-- Sequence identifier: @ReadID / 1 or 2 of pair
GACTCACGTAACTTTAAACTCTAACAGAAATATACTA…   <-- Sequence
+                                        <-- + (optionally lists the sequence identifier again)
CAEFGDG?BCGGGEEDGGHGHGDFHEIEGGDDDD…      <-- Quality String
```

### Base Quality Scores
The fourth line of each read is called the *quality string*.
Each symbol in the string is an encoding of the *quality score*, representing the inferred base call accuracy at that 
position in the read. 
The manufacturer of the sequencing instrument has performed calibration of quality score by sequencing many 
well-characterized samples from multiple organisms  and studying the correspondence between properties of the signal 
generated by the cluster being sequenced and the accuracy of the resulting base call.

The following two images explain this encoding.
The first image shows the mapping of the encoded quality score to the quality score:
<img src="../img/base_qual.png" width="500">

The second image shows the mapping of the quality score to the inferred base call accuracy:
<img src="../img/base_qual_table.png" width="700" >

Looking back at our sample read, we can see that the first base has an encoded quality score of `C`.
Using the first image above, we see that C encodes a quality of 34.
Using the second table, we see that the probability is < 1/1000 of that base being an error.
In the next section, we'll see how quality scores and other quality control metrics are used to evaluate the quality of 
a sequenced sample. 

More information on Quality scores from [Illumina](https://www.illumina.com/content/dam/illumina-marketing/documents/products/technotes/technote_understanding_quality_scores.pdf)
 
## FastQC
FastQC is widely used tool for both DNA and RNA sequencing data in order to evaluate the quality of the sequencing data.

To use, load the module: 
```bash
module load fastqc/0.11.8
```

To see the input options, type:
```bash
fastqc --help
```

Result:
```bash
fastqc --help

            FastQC - A high throughput sequence QC analysis tool

SYNOPSIS

	fastqc seqfile1 seqfile2 .. seqfileN

    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] 
           [-c contaminant file] seqfile1 .. seqfileN
...
```

FastQC is run on each FASTQ file separately in order to be sensitive to the variation in quality over lanes, samples, 
and paired-end files.

Since FastQC can run on multiple files at once, we'll use a wildcard `*` to indicate each file in the folder `raw_data`, 
and we specify that the output should be placed in the directory we created called `fastqc`:

```bash
cd ..
mkdir fastqc
fastqc raw_data/* -o fastqc
```

Result:
```markdown
Started analysis of na12878_1.fq
Approx 20% complete for na12878_1.fq
Approx 40% complete for na12878_1.fq
Approx 65% complete for na12878_1.fq
Approx 85% complete for na12878_1.fq
Analysis complete for na12878_1.fq
Started analysis of na12878_2.fq
Approx 20% complete for na12878_2.fq
Approx 40% complete for na12878_2.fq
Approx 65% complete for na12878_2.fq
Approx 85% complete for na12878_2.fq
Analysis complete for na12878_2.fq
```

To view the resulting files:
```
ls fastqc
```

The result shows an `html` file showing graphical results and a `zip` file containing the raw data for each input FASTQ file.
The easist way to view the result is to open the `html` files in a web browser.
```
na12878_1_fastqc.html  na12878_1_fastqc.zip  na12878_2_fastqc.html  na12878_2_fastqc.zip
```
## View results in the On Demand browser

To view the graphical results, return to the tab [ondemand.cluster.tufts.edu](ondemand.cluster.tufts.edu)

On the top menu bar choose `Files->Home Directory`

<img src="../img/od_files.png" width="400">

Navigate to the `fastqc` folder in course directory, e.g.: `/home/username/intro-to-ngs/fastqc/`
Right click on the file `na12878_1_fastqc.html` and select `Open in new tab`.

<img src="../img/od_new_tab.png" width="400">

The new tab that opens in the browser has the results of FastQC for the first reads in the sample.
We'll go through each plot.
Note that the plots shown are representative results for WES data of varying quality, rather than those generated 
on the course data.

### Per base sequence quality

The first plot shows the quality scores vs. position in the read, for all reads in the FASTQ file.

<img src="../img/fastqc_before_trim.png" width="400">

For each position a Box and Whisker type plot is drawn. The elements of the plot are as
follows:
- The central red line is the median value
- The yellow box represents the inter-quartile range (25-75%)
- The upper and lower whiskers represent the 10% and 90% points
- The blue line represents the mean quality

The background of the graph divides the y axis into very good quality calls
(green), calls of reasonable quality (orange), and calls of poor quality (red). 
It is typical for reads generated by Illumina platforms to show reduced quality at the ends of reads due to fragments in
 a cluster becoming out-of-sync ([Fuller et al 2009](http://arep.med.harvard.edu/pdf/Fuller_09.pdf)).


### Per sequence quality scores

The per sequence quality score report allows you to see if a subset of your sequences
have universally low quality values. It is often the case that a subset of sequences will
have universally poor quality, often because they are poorly imaged (on the edge of the
field of view etc), however these should represent only a small percentage of the total
sequences.

<img src="../img/fastqc_per_sequence_qual.png" width="400">

###Per base sequence content
Per Base Sequence Content plots out the proportion of each base position in a file for
which each of the four normal DNA bases has been called.

<img src="../img/fastqc_per_seq_content.png" width="400">

In a random library you would expect that there would be little to no difference between
the different bases of a sequence run, so the lines in this plot should run parallel with each
other. The relative amount of each base should reflect the overall amount of these bases
in your genome, but in any case they should not be hugely imbalanced from each other.
If you see strong biases which change in different bases then this usually indicates an
overrepresented sequence which is contaminating your library. A bias which is consistent
across all bases either indicates that the original library was sequence biased, or that
there was a systematic problem during the sequencing of the library.

###Per sequence GC content

This module measures the GC content across the whole length of each sequence in a file
and compares it to a modelled normal distribution of GC content.

<img src="../img/fastqc_gc.png" width="400">

In a normal random library you would expect to see a roughly normal distribution of GC
content where the central peak corresponds to the overall GC content of the underlying
genome. Since we don't know the the GC content of the genome the modal GC content is
calculated from the observed data and used to build a reference distribution.
An unusually shaped distribution could indicate a contaminated library or some other kinds
of biased subset. A normal distribution which is shifted indicates some systematic bias
which is independent of base position. If there is a systematic bias which creates a shifted
normal distribution then this won't be flagged as an error by the module since it doesn't
know what your genome's GC content should be.

###Per base N content

If a sequencer is unable to make a base call with sufficient confidence then it will normally
substitute an N rather than a conventional base] call
This module plots out the percentage of base calls at each position for which an N was
called.

<img src="../img/fastqc_n.png" width="400">

###Sequence Length Distribution

Some high throughput sequencers generate sequence fragments of uniform length, but
others can contain reads of wildly varying lengths. Even within uniform length libraries
some pipelines will trim sequences to remove poor quality base calls from the end.
This module generates a graph showing the distribution of fragment sizes in the file which
was analysed

<img src="../img/fastqc_length.png" width="400">

###Sequence Duplication Levels

In a diverse library most sequences will occur only once in the final set. A low level of
duplication may indicate a very high level of coverage of the target sequence, but a high
level of duplication is more likely to indicate some kind of enrichment bias (eg PCR over
amplification).

This module counts the degree of duplication for every sequence in the set and creates a
plot showing the relative number of sequences with different degrees of duplication.

<img src="../img/fastqc_dup.png" width="400">

###Overrepresented sequences

A normal high-throughput library will contain a diverse set of sequences, with no individual
sequence making up a tiny fraction of the whole. Finding that a single sequence is very
overrepresented in the set either means that it is highly biologically significant, or indicates
that the library is contaminated, or not as diverse as you expected.
This module lists all of the sequence which make up more than 0.1% of the total. To
conserve memory only sequences which appear in the first 200,000 sequences are
tracked to the end of the file. It is therefore possible that a sequence which is
overrepresented but doesn't appear at the start of the file for some reason could be
missed by this module.
For each overrepresented sequence the program will look for matches in a database of
common contaminants and will report the best hit it finds. Hits must be at least 20bp in
length and have no more than 1 mismatch. Finding a hit doesn't necessarily mean that this
is the source of the contamination, but may point you in the right direction. It's also worth
pointing out that many adapter sequences are very similar to each other so you may get a
hit reported which isn't technically correct, but which has very similar sequence to the
actual match.
Because the duplication detection requires an exact sequence match over the whole
length of the sequence any reads over 75bp in length are truncated to 50bp for the
purposes of this analysis. Even so, longer reads are more likely to contain sequencing
errors which will artificially increase the observed diversity and will tend to underrepresent
highly duplicated sequences.

###Adapter Content

This module looks for common adapters in the sequence.

<img src="../img/fastqc_adapter.png" width="400">

## Optional: Read trimming

In our `Per base sequence quality` we saw that the read quality dropped towards the end of the read.
This means that there are likely errors in the data.
In order to ensure alignment and variant calling are as accurate as possible, we can perform quality trimming of reads.

[Trim Galore](https://github.com/FelixKrueger/TrimGalore/blob/master/Docs/Trim_Galore_User_Guide.md) is a popular tool 
that in the default mode performs:
Quality trimming: Trims low quality bases from the 3' end of the read
Adapter trimming: Automatically detects and removes known Illumina adapters that may be present in the data

To perform trimming on the data, we first load the software:
```
module load anaconda/3
source activate /cluster/tufts/bio/tools/conda_envs/trim_galore/
```

To run:
```
mkdir trim
trim_galore -o trim raw_data/*
```

Result:
```
...
=== Summary ===

Total reads processed:                   4,652
Reads with adapters:                     1,606 (34.5%)
Reads written (passing filters):         4,652 (100.0%)

Total basepairs processed:       353,552 bp
Quality-trimmed:                  24,906 bp (7.0%)
Total written (filtered):        326,448 bp (92.3%)
...
```

Note that Trim Galore may trim adapters [even in the case where FastQC found no adapters](https://github.com/FelixKrueger/TrimGalore/issues/15).
This is because Trim Galore will remove partial adapters at the ends of reads.

The result after trimming is much improved:
<img src="../img/fastqc_dup.png" width="4

Explanations adapted from [https://dnacore.missouri.edu/PDF/FastQC_Manual.pdf][https://dnacore.missouri.edu/PDF/FastQC_Manual.pdf]
